### PR TITLE
#1946 Add tablet meta checkpoint mechanism

### PR DESCRIPTION
**Pull request by yiguolei**
1.  add rowset only update rowset meta manager
2. do meta checkpoint periodly
3. gc unused rowset metas

### CODE REVIEW

**Comment by yiguolei** [comment]
Add rowset should return success if rowset already exists. Currently, it returns version already exists but the callers will delete the rowset in this state.

File: be/src/olap/data_dir.cpp [hunk 1]
```diff
@@ -825,24 +825,18 @@ OLAPStatus DataDir::load() {
             }
         } else if (rowset_meta->rowset_state() == RowsetStatePB::VISIBLE 
             && rowset_meta->tablet_uid() == tablet->tablet_uid()) {
-            // add visible rowset to tablet, it maybe use in the future
-            // there should be only preparing rowset in meta env because visible 
-            // rowset is persist with tablet meta currently
-            OLAPStatus publish_status = tablet->add_inc_rowset(rowset);
+            OLAPStatus publish_status = OLAP_SUCCESS;
+            if (rowset_meta->is_singleton_delta()) {
+                publish_status = tablet->add_inc_rowset(rowset);
```
Comment (chaoyli) [inline be/src/olap/data_dir.cpp]: no need to add_inc_rowset

File: be/src/olap/data_dir.cpp [hunk 2]
```diff
@@ -825,24 +825,18 @@ OLAPStatus DataDir::load() {
...
+            } else {
+                publish_status = tablet->add_rowset(rowset, false);
+            }
             if (publish_status != OLAP_SUCCESS && publish_status != OLAP_ERR_PUSH_VERSION_ALREADY_EXIST) {
                 LOG(WARNING) << "add visilbe rowset to tablet failed rowset_id:" << rowset->rowset_id()
```
Comment (chaoyli) [inline be/src/olap/data_dir.cpp]: spell error, visible is right

File: be/src/olap/tablet.cpp [hunk 3]
```diff
@@ -240,13 +240,48 @@ OLAPStatus Tablet::deregister_tablet_from_dir() {
     return _data_dir->deregister_tablet(this);
 }
 
-OLAPStatus Tablet::add_rowset(RowsetSharedPtr rowset) {
+OLAPStatus Tablet::add_rowset(RowsetSharedPtr rowset, bool need_persist) {
     WriteLock wrlock(&_meta_lock);
+    // if the rowset already exist, should not return version already exist
+    // should return OLAP_SUCCESS
+    if (contains_rowset(rowset->rowset_id())) {
```
Comment (chaoyli) [inline be/src/olap/tablet.cpp]: if contains_rowset will be redundant with _check_added_rowset() ?

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by imay** [comment]
@yiguolei 

You should create an issue to describe this work's background. It will help other people easy to know this change list.

**Comment by yiguolei** [comment]
#1946

**Pull request merged**